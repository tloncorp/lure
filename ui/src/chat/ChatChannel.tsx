import React, { useEffect } from 'react';
import { Outlet, useNavigate, useParams } from 'react-router';
import { Helmet } from 'react-helmet';
import cn from 'classnames';
import ChatInput from '@/chat/ChatInput/ChatInput';
import ChatWindow from '@/chat/ChatWindow';
import Layout from '@/components/Layout/Layout';
import { ViewProps } from '@/types/groups';
import { useChatPerms, useChatState, useMessagesForChat } from '@/state/chat';
import {
  useRouteGroup,
  useVessel,
  useGroup,
  useChannel,
} from '@/state/groups/groups';
import ChannelHeader from '@/channels/ChannelHeader';
import useRecentChannel from '@/logic/useRecentChannel';
import { canReadChannel, canWriteChannel } from '@/logic/utils';

function ChatChannel({ title }: ViewProps) {
  const navigate = useNavigate();
  const { chShip, chName } = useParams();
  const chFlag = `${chShip}/${chName}`;
  const nest = `chat/${chFlag}`;
  const groupFlag = useRouteGroup();
  const { setRecentChannel } = useRecentChannel(groupFlag);

  useEffect(() => {
    useChatState.getState().initialize(chFlag);
    setRecentChannel(nest);
  }, [chFlag, nest, setRecentChannel]);

  const messages = useMessagesForChat(chFlag);
  const perms = useChatPerms(chFlag);
  const vessel = useVessel(groupFlag, window.our);
  const channel = useChannel(groupFlag, nest);
  const group = useGroup(groupFlag);
  const canWrite = canWriteChannel(perms, vessel, group?.bloc);
  const { sendMessage } = useChatState.getState();

  useEffect(() => {
    if (channel && !canReadChannel(channel, vessel, group?.bloc)) {
      navigate('../../activity');
      setRecentChannel('');
    }
  }, [group, channel, vessel, navigate, setRecentChannel]);

  return (
    <>
      <Layout
        className="flex-1 bg-white"
        header={<ChannelHeader flag={groupFlag} nest={nest} />}
        footer={
          <div className={cn(canWrite ? 'border-t-2 border-gray-50 p-4' : '')}>
            {canWrite ? (
              <ChatInput whom={chFlag} sendMessage={sendMessage} showReply />
            ) : null}
          </div>
        }
      >
        <Helmet>
          <title>
            {channel && group
              ? `${channel.meta.title} in ${group.meta.title} ${title}`
              : title}
          </title>
        </Helmet>
        <ChatWindow whom={chFlag} messages={messages} />
      </Layout>
      <Outlet />
    </>
  );
}

export default ChatChannel;
