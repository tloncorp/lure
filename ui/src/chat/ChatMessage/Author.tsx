import React from 'react';
import {
  makePrettyDayAndDateAndTime,
  makePrettyDayAndTime,
  makePrettyTime,
  useCopy,
} from '@/logic/utils';
import { useLocation } from 'react-router';
import { useModalNavigate } from '@/logic/routing';
import Avatar from '@/components/Avatar';
import ShipName from '@/components/ShipName';

interface AuthorProps {
  ship: string;
  date?: Date;
  timeOnly?: boolean;
  hideTime?: boolean;
}
export default function Author({
  ship,
  date,
  timeOnly,
  hideTime,
}: AuthorProps) {
  const location = useLocation();
  const { didCopy, doCopy } = useCopy(ship);
  const modalNavigate = useModalNavigate();
  const prettyTime = date ? makePrettyTime(date) : undefined;
  const prettyDayAndTime = date ? makePrettyDayAndTime(date) : undefined;
  const prettyDayAndDateAndTime = date
    ? makePrettyDayAndDateAndTime(date)
    : undefined;

  const handleProfileClick = () => {
    modalNavigate(`/profile/${ship}`, {
      state: { backgroundLocation: location },
    });
  };

  if (!date) {
    return (
      <div className="align-center group flex items-center space-x-3 py-1">
        <div onClick={handleProfileClick} className="shrink-0">
          <Avatar ship={ship} size="xs" className="cursor-pointer" />
        </div>
        <div
          onClick={doCopy}
          className="text-md shrink cursor-pointer font-semibold"
        >
          {didCopy ? (
            'Copied!'
          ) : (
            <ShipName
              name={ship}
              showAlias
              className="text-md font-semibold line-clamp-1"
            />
          )}
        </div>
      </div>
    );
  }

  return (
    <div className="align-center group flex items-center space-x-3 py-1">
      <div onClick={handleProfileClick} className="shrink-0">
        <Avatar ship={ship} size="xs" className="cursor-pointer" />
      </div>
      <div
        onClick={doCopy}
        className="text-md shrink cursor-pointer font-semibold"
      >
        {didCopy ? (
          'Copied!'
        ) : (
          <ShipName
            name={ship}
            showAlias
            className="text-md font-semibold line-clamp-1"
          />
        )}
      </div>

      {hideTime ? (
        <span className="hidden shrink-0 text-sm font-semibold text-gray-500 group-hover:block">
          {prettyDayAndTime}
        </span>
      ) : (
        <>
          <span className="hidden shrink-0 text-sm font-semibold text-gray-500 group-hover:block">
            {prettyDayAndDateAndTime}
          </span>
          <span className="block shrink-0 text-sm font-semibold text-gray-500 group-hover:hidden">
            {timeOnly ? prettyTime : prettyDayAndTime}
          </span>
        </>
      )}
    </div>
  );
}
