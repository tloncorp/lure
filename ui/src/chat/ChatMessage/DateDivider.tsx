import React, { Ref } from 'react';
import cn from 'classnames';
import { makePrettyDay, pluralize } from '@/logic/utils';

interface DateDividerProps {
  date: Date;
  unreadCount?: number;
}

function DateDividerComponent(
  { date, unreadCount }: DateDividerProps,
  ref: Ref<HTMLDivElement>
) {
  const prettyDay = makePrettyDay(date);

  return (
    <div ref={ref} className="flex w-full items-center py-4">
      <div
        className={cn(
          'h-[2px] w-8 rounded-sm',
          unreadCount ? 'bg-blue' : 'bg-gray-200'
        )}
      >
        &nbsp;
      </div>
      <span
        className={cn(
          'whitespace-nowrap px-3 font-semibold',
          unreadCount ? 'text-blue' : 'text-gray-400'
        )}
      >
        {prettyDay}
        {unreadCount ? (
          <>
            &nbsp;&bull;&nbsp;
            {unreadCount} new {pluralize('messsage', unreadCount)} below
          </>
        ) : null}
      </span>
      <div
        className={cn(
          'h-[2px] w-full rounded-sm',
          unreadCount ? 'bg-blue' : 'bg-gray-200'
        )}
      >
        &nbsp;
      </div>
    </div>
  );
}

const DateDivider = React.forwardRef(DateDividerComponent);

export default DateDivider;
