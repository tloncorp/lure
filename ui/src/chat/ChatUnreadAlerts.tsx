import React, { useCallback } from 'react';
import { format, isToday } from 'date-fns';
import { daToUnix, udToDec } from '@urbit/api';
import bigInt, { BigInteger } from 'big-integer';
import { VirtuosoHandle } from 'react-virtuoso';
import XIcon from '@/components/icons/XIcon';
import { pluralize } from '@/logic/utils';
import {
  useChatKeys,
  useChatState,
  useGetFirstUnreadID,
  useWrit,
} from '@/state/chat';
import { useChatInfo, useChatStore } from './useChatStore';

interface ChatUnreadAlertsProps {
  scrollerRef: React.RefObject<VirtuosoHandle>;
  whom: string;
}

export default function ChatUnreadAlerts({
  scrollerRef,
  whom,
}: ChatUnreadAlertsProps) {
  const chatInfo = useChatInfo(whom);
  const maybeWrit = useWrit(whom, chatInfo?.unread?.brief['read-id'] || '');
  let replyToId: BigInteger = bigInt(0);
  if (maybeWrit) {
    const [_, writ] = maybeWrit;
    const replyTo = writ.memo.replying || '';
    replyToId =
      replyTo !== ''
        ? bigInt(replyTo.split('/')[1].replaceAll('.', ''))
        : bigInt(0);
  }
  const markRead = useCallback(() => {
    useChatState.getState().markRead(whom);
    useChatStore.getState().read(whom);
  }, [whom]);

  // TODO: how to handle replies?
  const firstUnreadID = useGetFirstUnreadID(whom);
  const keys = useChatKeys({ whom, replying: false });
  const goToFirstUnread = useCallback(() => {
    if (!scrollerRef.current) {
      return;
    }

    if (replyToId !== bigInt(0)) {
      const idx = keys.findIndex((k) => k.greaterOrEquals(replyToId));
      if (idx === -1) {
        return;
      }

      scrollerRef.current.scrollToIndex({
        index: idx,
        align: 'start',
        behavior: 'auto',
      });
    }

    if (!firstUnreadID) {
      return;
    }

    const idx = keys.findIndex((k) => k.greaterOrEquals(firstUnreadID));
    if (idx === -1) {
      return;
    }

    scrollerRef.current.scrollToIndex({
      index: idx,
      align: 'start',
      behavior: 'auto',
    });
  }, [firstUnreadID, keys, scrollerRef, replyToId]);

  if (!chatInfo.unread || chatInfo.unread.seen) {
    return null;
  }

  const { brief } = chatInfo.unread;
  const readId = brief['read-id'];
  const udTime = readId
    ? daToUnix(bigInt(udToDec(readId.split('/')[1])))
    : null;
  const date = udTime ? new Date(udTime) : new Date();
  const since = isToday(date)
    ? `${format(date, 'HH:mm')} today`
    : format(date, 'LLLL d');

  const unreadMessage =
    brief &&
    `${brief.count} new ${pluralize('message', brief.count)} since ${since}`;

  if (!brief || brief?.count === 0) {
    return null;
  }

  return (
    <>
      <div className="absolute top-2 left-1/2 z-20 flex w-full -translate-x-1/2 flex-wrap items-center justify-center gap-2">
        <button
          className="button whitespace-nowrap bg-blue-soft text-sm text-blue dark:bg-blue-900 lg:text-base"
          onClick={goToFirstUnread}
        >
          <span className="whitespace-nowrap font-normal">
            {unreadMessage}&nbsp;&mdash;&nbsp;Click to View
          </span>
        </button>
        <button
          className="button whitespace-nowrap bg-blue-soft px-2 text-sm text-blue dark:bg-blue-900 lg:text-base"
          onClick={markRead}
        >
          <XIcon className="h-4 w-4" aria-label="Mark as Read" />
        </button>
      </div>
      <div />
    </>
  );
}
