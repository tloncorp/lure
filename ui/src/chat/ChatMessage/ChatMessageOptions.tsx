import React, { useCallback, useState } from 'react';
import { useNavigate } from 'react-router';
import { useCopy } from '@/logic/utils';
import { useAmAdmin, useRouteGroup } from '@/state/groups';
import { useChatState } from '@/state/chat';
import { ChatWrit } from '@/types/chat';
import IconButton from '@/components/IconButton';
import BubbleIcon from '@/components/icons/BubbleIcon';
import EllipsisIcon from '@/components/icons/EllipsisIcon';
import FaceIcon from '@/components/icons/FaceIcon';
import HashIcon from '@/components/icons/HashIcon';
import ShareIcon from '@/components/icons/ShareIcon';
import XIcon from '@/components/icons/XIcon';
import { useChatStore } from '@/chat/useChatStore';
import CopyIcon from '@/components/icons/CopyIcon';
import CheckIcon from '@/components/icons/CheckIcon';
import EmojiPicker from '@/components/EmojiPicker';

export default function ChatMessageOptions(props: {
  whom: string;
  writ: ChatWrit;
  hideReply?: boolean;
}) {
  const { whom, writ, hideReply } = props;
  const groupFlag = useRouteGroup();
  const isAdmin = useAmAdmin(groupFlag);
  const { didCopy, doCopy } = useCopy(
    `/1/chan/chat/${whom}/msg/${writ.seal.id}`
  );
  const [pickerOpen, setPickerOpen] = useState(false);

  const onDelete = () => {
    useChatState.getState().delMessage(whom, writ.seal.id);
  };

  const onCopy = useCallback(() => {
    doCopy();
  }, [doCopy]);

  const reply = useCallback(() => {
    useChatStore.getState().reply(whom, writ.seal.id);
  }, [writ, whom]);

  const onEmoji = useCallback(
    (emoji: { shortcodes: string }) => {
      useChatState.getState().addFeel(whom, writ.seal.id, emoji.shortcodes);
      setPickerOpen(false);
    },
    [whom, writ]
  );

  const openPicker = useCallback(() => setPickerOpen(true), [setPickerOpen]);

  const navigate = useNavigate();
  return (
    <div className="absolute right-2 -top-5 z-10 flex space-x-0.5 rounded-lg border border-gray-100 bg-white p-[1px] align-middle opacity-0 group-one-hover:opacity-100">
      <EmojiPicker
        open={pickerOpen}
        setOpen={setPickerOpen}
        onEmojiSelect={onEmoji}
      >
        <IconButton
          icon={<FaceIcon className="h-6 w-6 text-gray-400" />}
          label="React"
          showTooltip
          action={openPicker}
        />
      </EmojiPicker>
      {!writ.memo.replying && writ.memo.replying?.length !== 0 && !hideReply ? (
        <>
          {/*
          TODO: Add replies back in post-demo.
          <IconButton
            icon={<BubbleIcon className="h-6 w-6 text-gray-400" />}
            label="Reply"
            showTooltip
            action={reply}
          />

            */}
          <IconButton
            icon={<HashIcon className="h-6 w-6 text-gray-400" />}
            label="Start Thread"
            showTooltip
            action={() => navigate(`message/${writ.seal.id}`)}
          />
        </>
      ) : null}
      {groupFlag ? (
        <IconButton
          icon={
            didCopy ? (
              <CheckIcon className="h-6 w-6 text-gray-400" />
            ) : (
              <CopyIcon className="h-6 w-6 text-gray-400" />
            )
          }
          label="Copy"
          showTooltip
          action={onCopy}
        />
      ) : null}
      {/* <IconButton
        icon={<ShareIcon className="h-6 w-6 text-gray-400" />}
        label="Send to..."
        showTooltip
        action={() => console.log('send to..')}
      /> */}
      {isAdmin || window.our === writ.memo.author ? (
        <IconButton
          icon={<XIcon className="h-6 w-6 text-red" />}
          label="Delete"
          showTooltip
          action={onDelete}
        />
      ) : null}

      {/* <IconButton
        icon={<EllipsisIcon className="h-6 w-6 text-gray-400" />}
        label="More..."
        showTooltip
        action={() => console.log('More...')}
      /> */}
    </div>
  );
}
