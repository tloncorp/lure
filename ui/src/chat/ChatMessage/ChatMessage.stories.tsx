import { unixToDa } from '@urbit/api';
import React from 'react';
import { makeFakeChatWrit } from '../../mocks/chat';
import ChatMessage from './ChatMessage';

export default {
  component: 'ChatMessage',
  title: 'ChatMessage',
};

const writ = makeFakeChatWrit(1, '~finned-palmer', {
  block: [],
  inline: [{ bold: ['A bold test message'] }, 'with some more text'],
});
const time = unixToDa(writ.memo.sent);

export function Text() {
  return (
    <ChatMessage time={time} whom="~zod/test" writ={writ} newAuthor newDay />
  );
}
