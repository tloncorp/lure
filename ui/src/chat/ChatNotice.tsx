import React from 'react';
import AddPersonIcon from '@/components/icons/AddPersonIcon';
import ShipName from '@/components/ShipName';
import { ChatWrit } from '@/types/chat';
import DateDivider from './ChatMessage/DateDivider';

interface ChatNoticeProps {
  writ: ChatWrit;
  newDay?: Date;
}

const sp = '\u0020';
export default function ChatNotice({ writ, newDay }: ChatNoticeProps) {
  if (!('notice' in writ.memo.content)) {
    return null;
  }

  const { notice } = writ.memo.content;

  return (
    <div>
      {newDay ? <DateDivider date={newDay} /> : null}
      <div className="flex items-center space-x-3 py-2 pl-10 pr-2">
        <div className="flex items-center justify-center rounded bg-gray-50 p-0.5">
          <AddPersonIcon className="h-6 w-6 text-gray-600" />
        </div>
        <p className="font-semibold text-gray-400">
          {notice.pfix.length > 0 ? (
            <>
              {notice.pfix}
              {sp}
            </>
          ) : null}
          <ShipName name={writ.memo.author} showAlias />
          {notice.sfix.length > 0 ? (
            <>
              {sp}
              {notice.sfix}
            </>
          ) : null}
        </p>
      </div>
    </div>
  );
}
