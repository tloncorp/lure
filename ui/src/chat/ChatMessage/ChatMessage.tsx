/* eslint-disable react/no-unused-prop-types */
import React, { useCallback } from 'react';
import cn from 'classnames';
import _ from 'lodash';
import f from 'lodash/fp';
import { BigInteger } from 'big-integer';
import { daToUnix } from '@urbit/api';
import { format, formatDistanceToNow, formatRelative, isToday } from 'date-fns';
import { NavLink } from 'react-router-dom';
import { useInView } from 'react-intersection-observer';
import { ChatBrief, ChatWrit } from '@/types/chat';
import Author from '@/chat/ChatMessage/Author';
// eslint-disable-next-line import/no-cycle
import ChatContent from '@/chat/ChatContent/ChatContent';
import ChatReactions from '@/chat/ChatReactions/ChatReactions';
import DateDivider from '@/chat/ChatMessage/DateDivider';
import ChatMessageOptions from '@/chat/ChatMessage/ChatMessageOptions';
import {
  usePact,
  useChatState,
  useIsMessageDelivered,
  useIsMessagePosted,
  useWrit,
} from '@/state/chat';
import Avatar from '@/components/Avatar';
import DoubleCaretRightIcon from '@/components/icons/DoubleCaretRightIcon';
import UnreadIndicator from '@/components/Sidebar/UnreadIndicator';
import { useChatInfo, useChatStore } from '../useChatStore';

export interface ChatMessageProps {
  whom: string;
  time: BigInteger;
  writ: ChatWrit;
  newAuthor?: boolean;
  newDay?: boolean;
  hideReplies?: boolean;
  hideOptions?: boolean;
  isLast?: boolean;
  isLinked?: boolean;
  isScrolling?: boolean;
}

function briefMatches(brief: ChatBrief, id: string): boolean {
  return brief['read-id'] === id;
}

const ChatMessage = React.memo<
  ChatMessageProps & React.RefAttributes<HTMLDivElement>
>(
  React.forwardRef<HTMLDivElement, ChatMessageProps>(
    (
      {
        whom,
        time,
        writ,
        newAuthor = false,
        newDay = false,
        hideReplies = false,
        hideOptions = false,
        isLast = false,
        isLinked = false,
        isScrolling = false,
      }: ChatMessageProps,
      ref
    ) => {
      const { seal, memo } = writ;
      const chatInfo = useChatInfo(whom);
      const unread = chatInfo?.unread;
      const unreadId = unread?.brief['read-id'];
      const { ref: viewRef } = useInView({
        threshold: 1,
        onChange: useCallback(
          (inView) => {
            // if no tracked unread we don't need to take any action
            if (!unread) {
              return;
            }

            const { brief, seen } = unread;
            /* the first fire of this function
               which we don't to do anything with. */
            if (!inView && !seen) {
              return;
            }

            const {
              seen: markSeen,
              read,
              delayedRead,
            } = useChatStore.getState();
            const { markRead } = useChatState.getState();

            /* once the unseen marker comes into view we need to mark it
               as seen and start a timer to mark it read so it goes away.
               we ensure that the brief matches and hasn't changed before
               doing so. we don't want to accidentally clear unreads when
               the state has changed
            */
            if (inView && briefMatches(brief, writ.seal.id) && !seen) {
              markSeen(whom);
              delayedRead(whom, () => markRead(whom));
              return;
            }

            /* finally, if the marker transitions back to not being visible,
              we can assume the user is done and clear the unread. */
            if (!inView && unread && seen) {
              read(whom);
              markRead(whom);
            }
          },
          [unread, whom, writ.seal.id]
        ),
      });
      const isMessageDelivered = useIsMessageDelivered(seal.id);
      const isMessagePosted = useIsMessagePosted(seal.id);
      const isReplyOp = chatInfo?.replying === writ.seal.id;

      const unix = new Date(daToUnix(time));

      const pact = usePact(whom);

      const numReplies = seal.replied.length;
      const repliesContainsUnreadId = unreadId
        ? seal.replied.includes(unreadId)
        : false;
      const replyAuthors = _.flow(
        f.map((k: string) => {
          const t = pact.index[k];
          const mess = t ? pact.writs.get(t) : undefined;
          if (!mess) {
            return undefined;
          }
          return mess.memo.author;
        }),
        f.compact,
        f.uniq,
        f.take(3)
      )(seal.replied);
      const lastReply = seal.replied.length ? _.last(seal.replied)! : '';
      const lastReplyWrit = useWrit(whom, lastReply)!;
      const lastReplyTime = lastReplyWrit
        ? new Date(daToUnix(lastReplyWrit[0]))
        : new Date();

      return (
        <div
          ref={ref}
          className={cn('flex flex-col break-words', {
            'pt-2': newAuthor,
            'pb-2': isLast,
          })}
        >
          {unread && briefMatches(unread.brief, writ.seal.id) ? (
            <DateDivider
              date={unix}
              unreadCount={unread.brief.count}
              ref={viewRef}
            />
          ) : null}
          {newDay && !unread ? <DateDivider date={unix} /> : null}
          {newAuthor ? <Author ship={memo.author} date={unix} /> : null}
          <div className="group-one relative z-0 flex w-full">
            {hideOptions ? null : (
              <ChatMessageOptions
                hideReply={hideReplies}
                whom={whom}
                writ={writ}
              />
            )}
            <div className="-ml-1 mr-1 py-2 text-xs font-semibold text-gray-400 opacity-0 group-one-hover:opacity-100">
              {format(unix, 'HH:mm')}
            </div>
            <div className="wrap-anywhere flex w-full">
              <div
                className={cn(
                  'flex w-full grow flex-col space-y-2 rounded py-1 pl-3 pr-2 group-one-hover:bg-gray-50',
                  isReplyOp && 'bg-gray-50',
                  !isMessageDelivered && !isMessagePosted && 'text-gray-400',
                  isLinked && 'bg-blue-softer'
                )}
              >
                {'story' in memo.content ? (
                  <ChatContent
                    story={memo.content.story}
                    isScrolling={isScrolling}
                  />
                ) : null}
                {Object.keys(seal.feels).length > 0 && (
                  <ChatReactions seal={seal} whom={whom} />
                )}
                {numReplies > 0 && !hideReplies ? (
                  <NavLink
                    to={`message/${seal.id}`}
                    className={({ isActive }) =>
                      cn(
                        'rounded p-2 text-sm font-semibold text-blue',
                        isActive ? 'bg-gray-50' : ''
                      )
                    }
                  >
                    <div className="flex items-center space-x-2">
                      {repliesContainsUnreadId ? (
                        <UnreadIndicator
                          className="h-6 w-6 text-blue transition-opacity"
                          aria-label="Unread replies in this thread"
                        />
                      ) : null}
                      {replyAuthors.map((ship) => (
                        <Avatar key={ship} ship={ship} size="xs" />
                      ))}

                      <span>
                        {numReplies} {numReplies > 1 ? 'replies' : 'reply'}{' '}
                      </span>
                      <span className="text-gray-400">
                        Last reply{' '}
                        {isToday(lastReplyTime)
                          ? `${formatDistanceToNow(lastReplyTime)} ago`
                          : formatRelative(lastReplyTime, new Date())}
                      </span>
                    </div>
                  </NavLink>
                ) : null}
              </div>
              <div className="flex items-end rounded-r group-one-hover:bg-gray-50">
                {!isMessageDelivered && (
                  <DoubleCaretRightIcon
                    className="h-5 w-5"
                    primary={isMessagePosted ? 'text-black' : 'text-gray-200'}
                    secondary="text-gray-200"
                  />
                )}
              </div>
            </div>
          </div>
        </div>
      );
    }
  )
);

export default ChatMessage;
