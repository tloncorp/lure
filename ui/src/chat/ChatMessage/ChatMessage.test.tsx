import { describe, it, expect, beforeEach } from 'vitest';
import React from 'react';
import { MemoryRouter } from 'react-router';
import { unixToDa } from '@urbit/api';
import { TooltipProvider } from '@radix-ui/react-tooltip';
import { render } from '@testing-library/react';
import ChatMessage from './ChatMessage';
import { makeFakeChatWrit } from '../../mocks/chat';

describe('ChatMessage', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });
  it('renders as expected', () => {
    const date = new Date(2021, 1, 1, 13);
    const writ = makeFakeChatWrit(
      1,
      '~finned-palmer',
      {
        block: [],
        inline: [{ bold: ['A bold test message'] }, 'with some more text'],
      },
      undefined
    );
    const da = unixToDa(date.valueOf());
    const { asFragment } = render(
      <MemoryRouter>
        <TooltipProvider>
          <ChatMessage
            time={da}
            whom="~zod/test"
            writ={writ}
            newAuthor
            newDay
          />
        </TooltipProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
