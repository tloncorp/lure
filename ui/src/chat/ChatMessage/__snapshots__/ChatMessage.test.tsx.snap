// Vitest Snapshot v1

exports[`ChatMessage > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col break-words pt-2"
  >
    <div
      class="flex w-full items-center py-4"
    >
      <div
        class="h-[2px] w-8 rounded-sm bg-gray-200"
      >
         
      </div>
      <span
        class="whitespace-nowrap px-3 font-semibold text-gray-400"
      >
        February 1st
      </span>
      <div
        class="h-[2px] w-full rounded-sm bg-gray-200"
      >
         
      </div>
    </div>
    <div
      class="align-center group flex items-center space-x-3 py-1"
    >
      <div
        class="shrink-0"
      >
        <div
          class="relative flex flex-none items-center justify-center rounded bg-black w-6 h-6 rounded p-1.5 cursor-pointer"
          style="background-color: rgb(0, 0, 0);"
        >
          <svg
            class=""
            style="display: block; width: 12px; height: 12px;"
            version="1.1"
            viewBox="0 0 12 12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#000000"
              height="12px"
              width="12px"
              x="0px"
              y="0px"
            />
            <g
              transform="matrix(0.046875,0,0,0.046875,0,0)"
            >
              <g
                fill="none"
                transform="matrix(1,0,0,1,0,0)"
              >
                <rect
                  dataisgeon="true"
                  fill="white"
                  height="128"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                  width="128"
                />
              </g>
              <g
                fill="none"
                transform="matrix(1,0,0,1,128,0)"
              >
                <circle
                  cx="64"
                  cy="64"
                  dataisgeon="true"
                  fill="white"
                  r="64"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                />
              </g>
              <g
                fill="none"
                transform="matrix(1,0,0,1,0,128)"
              >
                <path
                  d="M0.0541 0C70.7217 0.0292317 128 57.3256 128 128C57.3177 128 0.0164917 70.7089 7.62806e-06 0.0305091C7.62851e-06 0.0203397 -4.44317e-10 0.01017 0 0H0.0541Z"
                  dataisgeon="true"
                  fill="white"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                />
              </g>
              <g
                fill="none"
                transform="matrix(1,0,0,1,128,128)"
              >
                <path
                  d="M64 128H0L5.59506e-06 0L64 5.59506e-06C99.3462 8.68512e-06 128 28.6538 128 64C128 99.3462 99.3462 128 64 128Z"
                  dataisgeon="true"
                  fill="white"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                />
              </g>
            </g>
          </svg>
        </div>
      </div>
      <div
        class="text-md shrink cursor-pointer font-semibold"
      >
        <span
          class="text-md font-semibold line-clamp-1"
        >
          <span
            aria-hidden="true"
          >
            ~
          </span>
          <span>
            finned
          </span>
          <span
            aria-hidden="true"
          >
            -
          </span>
          <span
            aria-hidden="false"
          >
            palmer
          </span>
        </span>
      </div>
      <span
        class="hidden shrink-0 text-sm font-semibold text-gray-500 group-hover:block"
      >
        February 1st, 2021 • 13:00
      </span>
      <span
        class="block shrink-0 text-sm font-semibold text-gray-500 group-hover:hidden"
      >
        February 1st • 13:00
      </span>
    </div>
    <div
      class="group-one relative z-0 flex w-full"
    >
      <div
        class="absolute right-2 -top-5 z-10 flex space-x-0.5 rounded-lg border border-gray-100 bg-white p-[1px] align-middle opacity-0 group-one-hover:opacity-100"
      >
        <button
          aria-controls="radix-0"
          aria-expanded="false"
          aria-haspopup="dialog"
          data-state="closed"
          type="button"
        >
          <div
            class="group-two cursor-pointer"
          />
        </button>
        <button
          aria-label="React"
          class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50 h-8 w-8"
          data-state="closed"
        >
          <svg
            class="h-6 w-6 text-gray-400"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fill-current"
              clip-rule="evenodd"
              d="M7 4h10c1.6569 0 3 1.34315 3 3v10c0 1.6569-1.3431 3-3 3H7c-1.65685 0-3-1.3431-3-3V7c0-1.65685 1.34315-3 3-3Zm10-2H7C4.23858 2 2 4.23858 2 7v10c0 2.7614 2.23858 5 5 5h10c2.7614 0 5-2.2386 5-5V7c0-2.76142-2.2386-5-5-5ZM8.5 8C7.67157 8 7 8.67157 7 9.5v1c0 .8284.67157 1.5 1.5 1.5s1.5-.6716 1.5-1.5v-1C10 8.67157 9.32843 8 8.5 8ZM14 9.5c0-.82843.6716-1.5 1.5-1.5s1.5.67157 1.5 1.5v1c0 .8284-.6716 1.5-1.5 1.5s-1.5-.6716-1.5-1.5v-1ZM13 15c0 .5523-.4477 1-1 1s-1-.4477-1-1 .4477-1 1-1 1 .4477 1 1Zm2 0c0 1.6569-1.3431 3-3 3s-3-1.3431-3-3 1.3431-3 3-3 3 1.3431 3 3Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="group-two cursor-pointer"
      >
        <button
          aria-label="Start Thread"
          class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50 h-8 w-8"
          data-state="closed"
        >
          <svg
            class="h-6 w-6 text-gray-400"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fill-current"
              clip-rule="evenodd"
              d="M10.124 3.00771c.5481.0685.9368.56829.8683 1.11631l-.4845 3.87596h3.9844l.5155-4.12403c.0685-.54802.5683-.93675 1.1163-.86824.5481.0685.9368.56829.8683 1.11631l-.4845 3.87596H19c.5523 0 1 .44772 1 1 0 .55229-.4477 1-1 1h-2.7422L15.7578 14H19c.5523 0 1 .4477 1 1s-.4477 1-1 1h-3.4922l-.5155 4.124c-.0685.548-.5683.9368-1.1163.8683-.5481-.0685-.9368-.5683-.8683-1.1164L13.4922 16H9.50778l-.5155 4.124c-.0685.548-.56829.9368-1.11631.8683s-.93675-.5683-.86825-1.1164L7.49222 16H5c-.55228 0-1-.4477-1-1s.44772-1 1-1h2.74222l.5-4.00002H5c-.55228 0-1-.44771-1-1 0-.55228.44772-1 1-1h3.49222l.5155-4.12403c.0685-.54802.56829-.93675 1.11628-.86824ZM13.7422 14l.5-4.00002h-3.9844L9.75778 14h3.98442Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="group-two cursor-pointer"
      >
        <button
          aria-label="Delete"
          class="flex h-8 w-8 items-center justify-center rounded group-two-hover:bg-gray-50 h-8 w-8"
          data-state="closed"
        >
          <svg
            class="h-6 w-6 text-red"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="stroke-current"
              d="m12 12 6-6m-6 6L6 6m6 6-6 6m6-6 6 6"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="-ml-1 mr-1 py-2 text-xs font-semibold text-gray-400 opacity-0 group-one-hover:opacity-100"
    >
      13:00
    </div>
    <div
      class="wrap-anywhere flex w-full"
    >
      <div
        class="flex w-full grow flex-col space-y-2 rounded py-1 pl-3 pr-2 group-one-hover:bg-gray-50"
      >
        <div
          class="leading-6"
        >
          <strong>
            <span>
              A bold test message
            </span>
          </strong>
          <span>
            with some more text
          </span>
        </div>
      </div>
      <div
        class="flex items-end rounded-r group-one-hover:bg-gray-50"
      />
    </div>
  </div>
</DocumentFragment>
`;
